registry.go

```diff
@@ -17,152 +17,247 @@ package schemaregistry
+			Subject: ss.Subject,
+			Version: ss.Version,
+			ID:      ss.ID,
+			Schema:  sr.Schema{}, // don't include schema in cache
+		})
 	}
+
+	return r, nil
 }
 
-func (r *SchemaRegistry) CreateSchema(ctx context.Context, subject string, schema sr.Schema) (sr.SubjectSchema, error) {
+//nolint:funlen // refactor at some point
+func (r *SchemaRegistry) CreateSchema(ctx context.Context, subject string, schema sr.Schema) (_ sr.SubjectSchema, err error) {
 	r.m.Lock()
 	defer r.m.Unlock()
 
-	fp := internal.Rabin([]byte(schema.Schema))
+	var rb rollback.R
+	defer rb.MustExecute()
+
+	fp := rabin.Bytes([]byte(schema.Schema))
 	id, ok := r.fingerprintIDCache[fp]
-	if ok {
-		// schema exists, see if subject matches
-		ss, err := r.findBySubjectID(subject, id)
-		if err == nil {
-			// schema exists for this subject, return it
-			return ss, nil
-		}
-		if !errors.Is(err, ErrSchemaNotFound) {
-			// unexpected error
-			return sr.SubjectSchema{}, err
-		}
-	} else {
+	if !ok {
 		// schema does not exist yet
 		id = r.nextID()
+		rb.AppendPure(r.revertID)
+	} else {
+		// check if a subject exists for this id
+		var ss sr.SubjectSchema
+		for _, s := range r.idSubjectSchemaCache[id] {
+			if s.Subject == subject && s.Version > ss.Version {
+				ss = s
+			}
+		}
+		if ss.ID != 0 {
+			// schema already exists for this subject, return it
+			ss.Schema = schema
+			return ss, nil
+		}
 	}
-	version := r.nextVersion(subject)
 
+	// we have to create a new SubjectSchema
 	ss := sr.SubjectSchema{
 		Subject: subject,
-		Version: version,
+		Version: r.nextVersion(subject),
 		ID:      id,
 		Schema:  schema,
 	}
 
-	r.schemas = append(r.schemas, ss)
+	txn := database.TransactionFromContext(ctx)
+	if txn != nil {
+		// the caller created a transaction, let them commit it
+		txn = nil
+	} else {
+		// the caller didn't create a transaction, we need to create one
+		// to ensure that the schema and subject schema are stored atomically
+		txn, ctx, err = r.db.NewTransaction(ctx, true)
+		if err != nil {
+			return sr.SubjectSchema{}, fmt.Errorf("failed to create transaction: %w", err)
+		}
+		defer txn.Discard()
+	}
+
+	err = r.schemaStore.Set(ctx, ss.ID, ss.Schema)
+	if err != nil {
+		return sr.SubjectSchema{}, err
+	}
+	err = r.subjectSchemaStore.Set(ctx, ss.Subject, ss.Version, ss)
+	if err != nil {
+		return sr.SubjectSchema{}, err
+	}
+
+	if txn != nil {
+		// commit the transaction if we created it
+		if err := txn.Commit(); err != nil {
+			return sr.SubjectSchema{}, fmt.Errorf("failed to commit transaction: %w", err)
+		}
+	}
+
+	// update caches
 	r.fingerprintIDCache[fp] = id
+	r.subjectVersionCache[subject] = append(r.subjectVersionCache[subject], ss.Version)
+	r.idSubjectSchemaCache[id] = append(r.idSubjectSchemaCache[id], sr.SubjectSchema{
+		Subject: ss.Subject,
+		Version: ss.Version,
+		ID:      ss.ID,
+		Schema:  sr.Schema{}, // don't include schema in cache
```

Comment:
I believe the field is initialized to `sr.Schema{}` by default, so I guess we have it here because of the comment?

Reply:
Yes, I wanted to make it explicit so when we look at this next time it doesn't look like an oversight.